### ISSUE TITLE
win/mingw64 build: mingw64.exe instead of cmd.exe ?

### ISSUE THREAD

**Issue by yuntan** [neovim/neovim#7059]
TL;DR Is it correct to use mingw64.exe instead of cmd.exe?

- neovim 4bcc70b2b5b02e3a6048b55a4390739570db3524
- Operating system/version: Windows 10 and [MSYS2 x86_64 20161025](http://www.msys2.org/)

I had followed instruction at https://github.com/neovim/neovim/wiki/Building-Neovim#windows--msys2 and got errors when executed `mingw32-make` at 'build' directory on cmd.exe.

```
[100%] Linking C executable ..\..\bin\nvim.exe
CMakeFiles\nvim.dir/objects.a(helpers.c.obj): In function `api_set_error':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/api/private/helpers.c:1027: undefined reference to `libintl_vsnprintf'
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/api/private/helpers.c:1033: undefined reference to `libintl_vsnprintf'
CMakeFiles\nvim.dir/objects.a(eval.c.obj): In function `fname_trans_sid':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/eval.c:6214: undefined reference to `libintl_snprintf'
CMakeFiles\nvim.dir/objects.a(eval.c.obj): In function `f_getmatches':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/eval.c:9885: undefined reference to `libintl_snprintf'
CMakeFiles\nvim.dir/objects.a(eval.c.obj): In function `f_setmatches':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/eval.c:14748: undefined reference to `libintl_snprintf'
CMakeFiles\nvim.dir/objects.a(eval.c.obj): In function `f_termopen':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/eval.c:16684: undefined reference to `libintl_snprintf'
CMakeFiles\nvim.dir/objects.a(eval.c.obj): In function `call_user_func':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/eval.c:21195: undefined reference to `libintl_snprintf'
CMakeFiles\nvim.dir/objects.a(eval.c.obj):C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/eval.c:22573: more undefined references to `libintl_snprintf' follow
CMakeFiles\nvim.dir/objects.a(ex_cmds2.c.obj): In function `get_locale_val':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/ex_cmds2.c:3411: undefined reference to `libintl_setlocale'
CMakeFiles\nvim.dir/objects.a(ex_cmds2.c.obj): In function `ex_compiler':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/ex_cmds2.c:2263: undefined reference to `libintl_snprintf'
CMakeFiles\nvim.dir/objects.a(ex_cmds2.c.obj): In function `ex_language':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/ex_cmds2.c:3554: undefined reference to `libintl_setlocale'
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/ex_cmds2.c:3568: undefined reference to `libintl_setlocale'
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/ex_cmds2.c:3571: undefined reference to `libintl_setlocale'
CMakeFiles\nvim.dir/objects.a(ex_docmd.c.obj): In function `eval_vars':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/ex_docmd.c:8547: undefined reference to `libintl_snprintf'
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/ex_docmd.c:8573: undefined reference to `libintl_snprintf'
CMakeFiles\nvim.dir/objects.a(ex_docmd.c.obj): In function `ex_terminal':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/ex_docmd.c:9822: undefined reference to `libintl_snprintf'
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/ex_docmd.c:9838: undefined reference to `libintl_snprintf'
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/ex_docmd.c:9844: undefined reference to `libintl_snprintf'
CMakeFiles\nvim.dir/objects.a(ex_getln.c.obj):C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/ex_getln.c:4430: more undefined references to `libintl_snprintf' follow
CMakeFiles\nvim.dir/objects.a(main.c.obj): In function `init_locale':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/main.c:678: undefined reference to `libintl_setlocale'
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/main.c:682: undefined reference to `libintl_setlocale'
CMakeFiles\nvim.dir/objects.a(main.c.obj): In function `usage':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/main.c:1882: undefined reference to `__printf__'
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/main.c:1883: undefined reference to `__printf__'
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/main.c:1884: undefined reference to `__printf__'
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/main.c:1885: undefined reference to `__printf__'
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/main.c:1886: undefined reference to `__printf__'
CMakeFiles\nvim.dir/objects.a(main.c.obj):C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/main.c:1887: more undefined references to `__printf__' follow
CMakeFiles\nvim.dir/objects.a(main.c.obj): In function `command_line_scan':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/main.c:1184: undefined reference to `libintl_snprintf'
CMakeFiles\nvim.dir/objects.a(mbyte.c.obj): In function `enc_locale':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/mbyte.c:1931: undefined reference to `libintl_setlocale'
CMakeFiles\nvim.dir/objects.a(message.c.obj): In function `get_emsg_source':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/message.c:395: undefined reference to `libintl_snprintf'
CMakeFiles\nvim.dir/objects.a(message.c.obj): In function `get_emsg_lnum':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/message.c:416: undefined reference to `libintl_snprintf'
CMakeFiles\nvim.dir/objects.a(message.c.obj): In function `msg_puts_printf':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/message.c:2034: undefined reference to `__printf__'
CMakeFiles\nvim.dir/objects.a(message.c.obj): In function `msg_outnum':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/message.c:1092: undefined reference to `libintl_snprintf'
CMakeFiles\nvim.dir/objects.a(channel.c.obj): In function `parse_msgpack':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/msgpack_rpc/channel.c:408: undefined reference to `libintl_snprintf'
CMakeFiles\nvim.dir/objects.a(channel.c.obj): In function `receive_msgpack':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/msgpack_rpc/channel.c:371: undefined reference to `libintl_snprintf'
CMakeFiles\nvim.dir/objects.a(channel.c.obj): In function `channel_write':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/msgpack_rpc/channel.c:562: undefined reference to `libintl_snprintf'
CMakeFiles\nvim.dir/objects.a(channel.c.obj): In function `handle_request':
C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/msgpack_rpc/channel.c:459: undefined reference to `libintl_snprintf'
CMakeFiles\nvim.dir/objects.a(server.c.obj):C:/Users/Yuuto/.local/src/github.com/neovim/neovim/src/nvim/msgpack_rpc/server.c:92: more undefined references to `libintl_snprintf' follow
collect2.exe: error: ld returned 1 exit status
mingw32-make[2]: *** [src\nvim\CMakeFiles\nvim.dir\build.make:4159: bin/nvim.exe] Error 1
mingw32-make[1]: *** [CMakeFiles\Makefile2:3092: src/nvim/CMakeFiles/nvim.dir/all] Error 2
mingw32-make: *** [Makefile:151: all] Error 2
```

Note that I had updated packages by `pacman -Syu` before installing required packages.

After several try and error, I noticed that it goes well (linking to nvim.exe finished successfully) on mingw64.exe not on cmd.exe.

```
mkdir build; cd build
CC=gcc cmake -G "MinGW Makefiles" ..
mingw32-make
```

**Comment by justinmk** [comment]
Never heard of `mingw64.exe`. Is that a shell, like cmd.exe?

**Comment by yuntan** [comment]
I found it at "C:\msys64\mingw64.exe"

**Comment by justinmk** [comment]
I see that `mingw64.exe` is a mintty terminal which launches a msys2 bash shell.

There was some reason we explicitly use cmd.exe instead of msys2 environment. We do this in `ci/build.bat`. I don't remember why, maybe @jamessan @equalsraf can say.

**Comment by justinmk** [comment]
Side note: it could be really helpful if someone generalized `ci/build.bat` to be usable for developers, not just AppVeyor.

**Comment by equalsraf** [comment]
I think cmd.exe was used because sh.exe must not be in PATH (if using -G "MinGW Makefiles" https://cmake.org/Wiki/CMake_MinGW_Compiler_Issues). Maybe this changed in cmake?

**Comment by equalsraf** [comment]
Tested here. I still get an error when trying to cmake -G "MinGW Makefiles" about sh.exe

**Comment by yuntan** [comment]
I remember that I experienced the same inconsistent behavior of cmake. When I called `cmake -G "MinGW Makefiles"` at first time, I got error about sh.exe, but second time I called, it passes.

I found that the first problem (linking failed due to undefined reference to ...) is my environment dependent problem. Having different version of libintl-8.dll in my PATH was causing it. `set PATH=c:\msys64\mingw64\bin;%PATH%` instead of `set PATH=%PATH%;c:\msys64\mingw64\bin` solved it.

**Comment by equalsraf** [comment]
> Side note: it could be really helpful if someone generalized ci/build.bat to be usable for developers, not just AppVeyor.

@justinmk I have some thoughts on that but no clean solution yet. If someone wants to give it a whirl, check my notes https://gist.github.com/equalsraf/8710cd7324df0f67d047a99418356317

**Comment by janlazo** [comment]
I prefer to keep using a batchfile to leave the option open for msvc (for the unit tests), at least in Appveyor, because it requires calling a batchfile to setup `cl.exe`.
It shouldn't be difficult to break up `ci/build.bat` into smaller scripts, at least to isolate Appveyor-specific configurations, and run them in a child shell if necessary. I do this for ruby and node batchfiles to install the providers.
This should be useful for legacy tests in `src/nvim/testdir` where I currently append msys2's bin/ so make can use sh.exe.

**Comment by justinmk** [comment]
This is not actionable (except "make things better") and there was no followup. https://github.com/neovim/neovim/pull/8084 allows building with MSVC as an alternative.
